### ISSUE TITLE
[Bug][CPU EP][Apple Silicon] Identical SGEMM input becomes nondeterministic with KleidiAI in ONNX Runtime 1.27.0

### ISSUE THREAD

**Issue by JD3Lasers** [microsoft/onnxruntime#31569]
### Describe the issue

ONNX Runtime 1.27.0's CPU Execution Provider produces nondeterministic output
for identical input on Apple Silicon when the Arm KleidiAI SGEMM path is
enabled. The same all-zero `float32` tensor is submitted three times through
one unchanged `InferenceSession`; the third `beat` output differs from the
first two.

Applying the documented session configuration
`mlas.disable_kleidiai=1` restores bit-identical output. ONNX Runtime 1.26.0
also remains bit-identical with default settings, so this appears to be a
1.27.0 regression.

Expected: identical input submitted repeatedly through an unchanged inference
session produces identical output.

### To reproduce

The standalone reproducer, dependency pins, model provenance, observed
digests, and downloadable ONNX model are public here:

- Repository: https://github.com/JD3Lasers/onnxruntime-kleidiai-reproducer
- Script: https://raw.githubusercontent.com/JD3Lasers/onnxruntime-kleidiai-reproducer/main/reproduce_onnxruntime_kleidiai.py
- Model: https://github.com/JD3Lasers/onnxruntime-kleidiai-reproducer/releases/download/v1.0.0/beatthis-final0.onnx
- Manifest: https://raw.githubusercontent.com/JD3Lasers/onnxruntime-kleidiai-reproducer/main/beatthis_onnx_model_manifest.json

```bash
git clone https://github.com/JD3Lasers/onnxruntime-kleidiai-reproducer.git
cd onnxruntime-kleidiai-reproducer
python -m venv .venv
.venv/bin/pip install -r requirements.txt
.venv/bin/python reproduce_onnxruntime_kleidiai.py /path/to/beatthis-final0.onnx
```

The command creates one CPU EP session and submits the same all-zero
`float32 (1, 1500, 128)` tensor three times. It repeats the experiment in a
second session configured with `mlas.disable_kleidiai=1`.

Observed results:

| ONNX Runtime | Session | All three equal | Changed output elements | Maximum absolute difference |
| --- | --- | --- | ---: | ---: |
| 1.27.0 | default | no | 1,500 | 0.0026569366455078125 |
| 1.27.0 | `mlas.disable_kleidiai=1` | yes | 0 | 0 |
| 1.26.0 | default | yes | 0 | 0 |
| 1.26.0 | `mlas.disable_kleidiai=1` | yes | 0 | 0 |

On 1.27.0 with defaults, the first two output SHA-256 digests are
`e7298cadec17c78236f1dd667f8010b8425949ebaa4a22e007fbe8f336ef9f0a`;
the third is
`b889acda3c5687c629238f03f69cf0bfe7cc7c760fdbd9c64ad23f0a2e436474`.
The disabled control is stable at
`1e2847f203fcb6b5e53c7973b929e387c4ca89d2ad7262736a7be550b1377a4c`.

The ONNX model SHA-256 is
`e0024b9af8f9a0da5e2541e89ec92cfca82530aaaa33a9692d6f22626f0ebaf7`.

microsoft/onnxruntime#28654 appears distinct: it affected 1.26.0 and reported
cross-model convolution scratchpad contamination, while this reproducer is
stable on 1.26.0 and regresses on 1.27.0 with a KleidiAI-specific control.

### Urgency

The application has a safe session-level opt-out, so this is not currently
blocking a release. The report is intended to identify the regression and
determine when the opt-out can be removed safely.

### System information

- Platform: Mac
- OS version: macOS 26.5.2 (25F84)
- Installation: released `onnxruntime` CPU package from PyPI
- ONNX Runtime versions: 1.27.0 affected; 1.26.0 control
- API: Python 3.14.6
- Architecture: ARM64, Apple M5 Pro
- Execution Provider: Default CPU
- NumPy: 2.4.6

**Comment by hariharans29** [comment]
Thanks for this bug report. I will bring it to the attention of relevant folks.

**Comment by hariharans29** [comment]
I just noticed that you are on ORT 1.27. Since our latest is 1.28 - can you please check if this repros there ?

**Comment by JD3Lasers** [comment]
Tested ONNX Runtime 1.28.0 as requested on the same reproducer host and model.

Environment: macOS 26.5.2 arm64 (Apple M5 Pro), Python 3.14.6, NumPy 2.4.6, model SHA-256 `e0024b9af8f9a0da5e2541e89ec92cfca82530aaaa33a9692d6f22626f0ebaf7`.

Results:
- 1.27.0 default reproduced the original failure in 3/3 fresh processes: the third replay changed all 1,500 output elements, maximum absolute difference `0.0026569366455078125`, with the same digests from the report.
- 1.28.0 default passed in 5/5 fresh processes: all three replays were bit-identical in every process, with zero changed elements and zero maximum difference. The stable default digest was `8600044ba640de2a83e911ba9200b9fbc59f36cbf512d804fe50ee44cc14ccd7`.
- The `mlas.disable_kleidiai=1` control remained stable on both versions.

We also ran the application packaged-model determinism and leading-silence correctness gates under 1.28.0 with KleidiAI enabled; both passed. Default and disabled 1.28.0 logits are not bit-identical, but the difference is small on the reproducer input: 1,378/1,500 values differ, maximum absolute difference `1.9073486328125e-05`, mean absolute difference `3.97555049858056e-06`.

This issue does not reproduce for us on the released 1.28.0 macOS arm64 wheel.

**Comment by JonathanC-ARM** [comment]
Hi @hariharans29 @JD3Lasers, I did a bit of checking on this to be sure. 

The offending code was an incorrect relaxation of padding requirements for IGemm KleidiAI Kernels, which was subsequently fixed as part of https://github.com/microsoft/onnxruntime/pull/28571. As you had stated this is no longer reproducible in 1.28 ort whl, and the fix would be in any subsequent whl's built from main since Jun 19th 

Thanks for the detailed report, it made checking things easy on my end and Apologies for the inconvenience.
